File: src/units/RobotUnit.scala
package units

import battlecode.common.{GameActionException, RobotController}

/**
  * Base trait for units.
  */
trait RobotUnit {

  @throws(classOf[GameActionException])
  def run()(implicit rc: RobotController): Unit =
    while (true)
      runStep()

  @throws(classOf[GameActionException])
  def runStep()(implicit rc: RobotController): Unit = ???
}

File: src/units/RobotUnit.java
package units;

import battlecode.common.GameActionException;

/**
 * Base trait for units.
 */
public interface RobotUnit {

    /*
     Fine-grained stages
      */
    default void prerun() {}
    default void postrun() {}

    /*
     General
      */
    default void run() {
        while (true)
            try {
                prerun();
                runStep();
                postrun();
            }
            catch (GameActionException e) {
                System.out.println("Game Exception");
                e.printStackTrace();
            }
    }

    void runStep() throws GameActionException;
}
